mlir/test/Integration/Dialect/ControlFlow/test-region-simplification.mlir

```diff
@@ -0,0 +1,49 @@
+// Baseline check
+// RUN: mlir-opt %s --convert-func-to-llvm --convert-cf-to-llvm | \
+// RUN: mlir-cpu-runner -e nested_loop --entry-point-result=i32 | FileCheck %s
+
+// Region simplification check
+// RUN: mlir-opt %s \
+// RUN: --canonicalize='enable-patterns=AnyPattern region-simplify=aggressive' \
+// RUN: --convert-func-to-llvm --convert-cf-to-llvm | mlir-cpu-runner \
+// RUN: -e nested_loop --entry-point-result=i32 | FileCheck %s
+
+func.func @nested_loop() -> i32 {
+  %c3_i64 = arith.constant 3 : i64
+  %c2_i64 = arith.constant 2 : i64
+  %c0_i64 = arith.constant 0 : i64
+  %c1_i64 = arith.constant 1 : i64
+  %c1_i32 = arith.constant 1 : i32
+  %c0_i32 = arith.constant 0 : i32
+  cf.br ^bb1(%c0_i32, %c0_i64 : i32, i64)
+^bb1(%0: i32, %1: i64):  // 2 preds: ^bb0, ^bb8
+  %2 = arith.cmpi ult, %1, %c2_i64 : i64
+  cf.cond_br %2, ^bb2(%0, %1 : i32, i64), ^bb9(%0, %1 : i32, i64)
+^bb2(%3: i32, %4: i64):  // pred: ^bb1
+  %5 = arith.addi %4, %c1_i64 : i64
+  cf.br ^bb3(%3, %5 : i32, i64)
+^bb3(%6: i32, %7: i64):  // 2 preds: ^bb2, ^bb5
+  %8 = arith.cmpi ult, %7, %c3_i64 : i64
+  cf.cond_br %8, ^bb4(%6, %7 : i32, i64), ^bb6(%6, %7 : i32, i64)
+^bb4(%9: i32, %10: i64):  // pred: ^bb3
+  %11 = arith.addi %9, %c1_i32 : i32
+  cf.br ^bb5(%11, %10 : i32, i64)
+^bb5(%12: i32, %13: i64):  // pred: ^bb4
+  %14 = arith.addi %13, %c1_i64 : i64
+  cf.br ^bb3(%12, %14 : i32, i64)
+^bb6(%15: i32, %16: i64):  // pred: ^bb3
+  cf.br ^bb7
+^bb7:  // pred: ^bb6
+  cf.br ^bb8(%15, %4 : i32, i64)
+^bb8(%17: i32, %18: i64):  // pred: ^bb7
+  %19 = arith.addi %18, %c1_i64 : i64
+  cf.br ^bb1(%17, %19 : i32, i64)
+^bb9(%20: i32, %21: i64):  // pred: ^bb1
+  cf.br ^bb10
+^bb10:  // pred: ^bb9
+  return %20 : i32
+}
+
+// If region simplification behaves correctly (by NOT merging ^bb2 and ^bb5),
+// this will be 3.
+// CHECK: 3
```

Comment:
Can you please replace the integration test by a unit-test: it should go in `mlir/test/Transforms/canonicalize-block-merge.mlir`. Thanks!

Reply:
OK, I can do that.

The command at the top of that file (`mlir-opt -allow-unregistered-dialect %s -pass-pipeline='builtin.module(func.func(canonicalize{region-simplify=aggressive}))'`) won't exercise the new code because if one runs canonicalization before region simplification, the problem is averted. The integration test that I ran intentionally added `enable-patterns=AnyPattern` to skip the typical canonicalization processing that masks the problem.

(I realize this is a bit convoluted, but the `mlir-opt` command line that I'm using is essentially trying to mimic what happens when MLIR code calls `applyPatternsAndFoldGreedily()`, hence the need to skip the regular canonicalizaiton.)

So I will need to add an additional `mlir-opt` command to the top of that file with a custom `FileCheck --check-prefix=PATTERNS`. Does that sound reasonable to you?

Also, I think the unit test would need to remain pretty much as complicated as the IR for the integration test in order to exercise this bugfix. Is it ok to simply copy the IR from the `test-region-simplification.mlir` and put it into the new unit test file? Or were you wanting a simpler test case? Any attempts to simplify it to a smaller number of blocks leads has thus far led to `error: operand #1 does not dominate this use`.

Reply:
> So I will need to add an additional mlir-opt command to the top of that file with a custom FileCheck --check-prefix=PATTERNS. Does that sound reasonable to you?

Probably worth just adding a new file next to this one.

Test case should be the simplest, it's not clear to me about the source of the `error: operand #1 does not dominate this use`?  Is this the verifier on the input file or after the pass runs? If the input then that seems like an issue in the way you reduce it isn't it?

Reply:
Yes, I've discussed this a bit with @schweitzpgi. For a fully correct IR that truly demonstrates the bug, I believe it will require 6 basic blocks.
![image](https://github.com/user-attachments/assets/5b8f7885-679d-41fa-888f-eb517334bb79)

In this hypothetical case, bb2 and bb4 are identical and could be merged *except* if bb5 uses a block argument from bb2.

The scenario is fairly complicated, so that's why I originally opted for an integration test, because it definitively shows that the problem exists (before the bugfix).

Every attempt I've made to simplify it down to something smaller causes one of two things to happen: 1) the input validation fails because I've simplified it into an invalid test case, or 2) the other IR simplifications on the *output* makes it difficult to discern whether the bug has truly been resolved.

For example, regarding the first path (where input validation fails because I've simplified it into an invalid test case), if I construct a test case like this:
```mlir
func.func @nomerge2(%arg0: i32, %i: i32) -> i32 {
  %c1_i32 = arith.constant 1 : i32
  %icmp = arith.cmpi slt, %i, %arg0 : i32
  cf.cond_br %icmp, ^bb2(%arg0 : i32), ^bb3(%c1_i32 : i32)

^bb2(%bbarg2 : i32):
  %ip1 = arith.addi %bbarg2, %c1_i32 : i32
  cf.br ^bb4(%ip1 : i32)

^bb3(%bbarg3 : i32):
  %jp1 = arith.addi %bbarg3, %c1_i32 : i32
  cf.br ^bb4(%jp1 : i32)

^bb4(%j: i32):
  %jcmp = arith.cmpi slt, %j, %bbarg2 : i32
  return %bbarg2 : i32
}
```
Then the compiler correctly reports this error message and will not run the pass:
```
error: operand #1 does not dominate this use
  %jcmp = arith.cmpi slt, %j, %bbarg2 : i32
```

So if we have to construct valid IR for the input test case, then I think we will be left with a fairly complicated (and therefore difficult to interpret) test. So the integration test was an attempt to mitigate that by simply having a single numerical output tell you whether the bug has been fixed or not.

Reply:
FWIW, here is the smallest example I can construct that shows the problem in `main` and demonstrates that it is fixed with the changes in this PR:
```mlir
func.func @nested_loop() -> i32 {
  %c0_i32 = arith.constant 0 : i32
  %c1_i32 = arith.constant 1 : i32
  %c2_i32 = arith.constant 2 : i32
  %c3_i32 = arith.constant 3 : i32
  cf.br ^bb1(%c0_i32, %c0_i32 : i32, i32)
^bb1(%0: i32, %1: i32):
  %2 = arith.cmpi ult, %1, %c2_i32 : i32
  cf.cond_br %2, ^bb2(%0, %1 : i32, i32), ^bb7(%0 : i32)
^bb2(%3: i32, %4: i32):
  %5 = arith.addi %4, %c1_i32 : i32
  cf.br ^bb3(%3, %5 : i32, i32)
^bb3(%6: i32, %7: i32):
  %8 = arith.cmpi ult, %7, %c3_i32 : i32
  cf.cond_br %8, ^bb4(%6, %7 : i32, i32), ^bb6(%6, %4 : i32, i32)
^bb4(%9: i32, %10: i32):
  %11 = arith.addi %9, %c1_i32 : i32
  cf.br ^bb5(%11, %10 : i32, i32)
^bb5(%12: i32, %13: i32):
  %14 = arith.addi %13, %c1_i32 : i32
  cf.br ^bb3(%12, %14 : i32, i32)
^bb6(%15: i32, %16: i32):
  %17 = arith.addi %16, %c1_i32 : i32
  cf.br ^bb1(%15, %17 : i32, i32)
^bb7(%18: i32):
  return %18 : i32
}
```
If you place the above into `test.mlir` and run with the following command, the latest on `main` will return `2` while the fixed version returns `3`.
```bash
$ build/bin/mlir-opt --canonicalize='enable-patterns=AnyPattern region-simplify=aggressive' --convert-func-to-llvm --convert-cf-to-llvm test.mlir | build/bin/mlir-cpu-runner -e nested_loop --entry-point-result=i32
```

In this example, the original problem reported by the original issue occurs when `^bb2` and `^bb5` are merged.

Reply:
If I were to turn this into a unit test, the output (*) of the above test is difficult to prove that the problem is fixed, which I fear will make a unit test difficult to maintain going forward.

(*) Here is the output of the aforementioned `test.mlir` after running through the pass, and `test.mlir` is the simplest test case that I can make.
```mlir
$ mlir-opt --canonicalize='enable-patterns=AnyPattern region-simplify=aggressive' test.mlir
module {
  func.func @nested_loop() -> i32 {
    %c0_i32 = arith.constant 0 : i32
    %c1_i32 = arith.constant 1 : i32
    %c2_i32 = arith.constant 2 : i32
    %c3_i32 = arith.constant 3 : i32
    cf.br ^bb1(%c0_i32, %c0_i32 : i32, i32)
  ^bb1(%0: i32, %1: i32):  // 2 preds: ^bb0, ^bb5
    %2 = arith.cmpi ult, %1, %c2_i32 : i32
    cf.cond_br %2, ^bb2(%1, %0 : i32, i32), ^bb6
  ^bb2(%3: i32, %4: i32):  // 2 preds: ^bb1, ^bb4
    %5 = arith.addi %3, %c1_i32 : i32
    cf.br ^bb3
  ^bb3:  // pred: ^bb2
    %6 = arith.cmpi ult, %5, %c3_i32 : i32
    cf.cond_br %6, ^bb4, ^bb5
  ^bb4:  // pred: ^bb3
    %7 = arith.addi %4, %c1_i32 : i32
    cf.br ^bb2(%5, %7 : i32, i32)
  ^bb5:  // pred: ^bb3
    %8 = arith.addi %1, %c1_i32 : i32
    cf.br ^bb1(%4, %8 : i32, i32)
  ^bb6:  // pred: ^bb1
    return %0 : i32
  }
}
```

Note that after `^bb2`, this IR bears little resemblance to the input `test.mlir`, so it's difficult to determine whether or not `^bb2` and `^bb5` were merged with any `CHECK` rules.

In fact, here are the differences between bad (`main`) and good (this PR) when running `mlir-opt --canonicalize='enable-patterns=AnyPattern region-simplify=aggressive' test.mlir`:

![image](https://github.com/user-attachments/assets/3d0589f6-4651-4349-92ba-7f0f175d6f85)

If I were to write `CHECK` rules to look for the patterns on the right and `CHECK-NOT` the patterns on the left, it would be very difficult to maintain those going forward.

In summary, the other downstream IR changes from the pass make it very difficult to make this into a good and maintainable unit test in my opinion.

So I would like to propose that rather than making a unit test, I update the integration test with the simpler version of IR (i.e. `test.mlir` in this thread). Does that sound reasonable?

Reply:
OK, I replaced the integration test with a unit test here: 5b3ee9ffc8e5e5a8d8113316d328a2d6b8eb01d5. Please let me know if that's what you had in mind. (Thanks for your offline help, @joker-eph.)